supercollider/supercollider
# SuperCollider projects V2

This is a revised version of PR #1978. I've updated the code taking into account the changes added in #2861. The code has also been substantially changed with some features added.

## Overview
A supercollider project is an sclang config file which has an extra field "project: true". When this field is present sclang behaves differently:

- Relative paths in the config file are resolved relative to the directory containing the config file.
- Quarks download to the directory containing the config file.
- If present in the same directory as the config file a startup.scd is run at startup. The global startup.scd is not run.

This allows:
- Having self-contained projects which can be shared with other users by just sharing the directory containing the project file.
- Creating sclang standalones, that is, applications based on on supercollider and with a sclang Qt gui which can be installed and shared. (Currently the part of doing a system install or having a MacOS bundle is not implemented).




## Implementation notes:

### general

- It is recommended that when using a project to also activate the exclude default paths option, but this was not hardcoded for flexibility.
- The extension for project config files is by default .scproj

### sclang

- excluding default paths was removed from sclang command line options and moved to the config file. This introduces a change in the command line options of sclang, removing the -a option. The rationale is that with -a a lang file has to be used anyhow otherwise the class library will not compile as it doesn't find an SCClassLibrary, therefore might as well make that an option which is settable in the language config file which also makes it easier to manage. This also changes the header API to the state it was in before pull-request #1021, it is unfortunate going back and forth, but I do think it is better as it is in this pull-request.
- readLibraryConfigYAML: code was simplified abstracting repeated code using lambdas. In the future more simplification could be achieved by sharing this code between the ide and sclang, although there are issues when templated code interacts with lambdas in C++11.
- PyrLexer: gResourceDirectory, gSystemExtensionDirectory, gUserExtensionDirectory are initialized once in initPassOne(). In passOne_ProcessDir the corresponding strings are substituted by the actual paths and relative paths are converted to absolute (using the path of the language config file).
- Platform.sc: in project mode, if a start.scd file is present in the same directory as the config file it is used at startup instead of global start.scd file (which in any case is not used if exclude default paths is activated).

### IDE

- An option was added to store the language config in the sessions (disabled by default). This can be changed in the settings dialog. It allows having a different project or config file associated with each session.
- Changing from storing lang in session or settings will update dialog and guis while still in the dialog (without doing apply). The method void Dialog::switchSclangPage(bool useLanguageConfigFromSession) is used in the preference panes because that choice is still not commited to the QSettings object.
- Settings wraps a QSettings and session inherits from QSettings, but they both follow the same interface. This was formalized using an abstract class SettingsInterface from which both inherit. This contributes to very significant simplification of code dealing with fetching and writting the langauge config in either settings or session. This also allowed removing some templates from a couple of methods.
- A filesystem tree view was added to show the contents of the project directory.
- sclang_page_generic.hpp: both project and non project pages are very similar, therefore a class was created which includes common functionality. This class is templated on the type of the UI class which corresponds to the ui form.
- verifyProjectStatusMatchesConfigFile(): At startup and when changing sessions the config files (in the global config or in the session, depending on the current option) are checked to see detect if they of project type.
- When a project is open this is reflected in the title bar, sclang dialog page, Quarks gui and project file tree.
- When closing a project the selected config file goes back to the previous non-project config file (which is saved in the settings).

### Quarks

- In Quarks.sc relative paths are converted to absolute paths, in sclang c++ code they are converted from absolute to relative paths when part of a project. When adding a quark in project mode the absolute path will be passed to sclang, it is then converted to a relative path. This ensures that when a project directory is moved to a different path, the included and excluded folders which are part of the project are still found.

### Standalone

An sclang standalone consists of
- A project file.
- sclang and scsynth binaries
- plugins
- synthdefs
- class files
- quarks
- start.scd: for startup code
- start: a start script just calls 'sclang -l configfile.yaml'.

There is a template (without binaries) available at https://github.com/miguel-negrao/scStandalone.

## Testing

- There is test project available at https://github.com/miguel-negrao/TestSCProject

How to test this PR:
- create a new project, add paths in settings, install quarks.
- close project
- activate in the settings dialog the option for saving config file in sessions.
- create different sessions, open a project in some sessions and use a different config file in others. Change sessions.
- open project, close ide. Change project to false in the text file. Open ide again, confirm that it doesn't open as project and the current config file is last non-project config file.
